File: tests/drivers/node/driver_compat.js
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// PROD-8 ioredis compatibility driver (issue #158).
//
// Runs the real `ioredis` client against a live IronCache, in TWO modes:
//
//   * single-node (new Redis()): core data-type ops, pipelining, MULTI/EXEC, pub/sub, and a RESP3
//     HELLO probe.
//   * cluster (new Redis.Cluster([...])): topology DISCOVERY (ioredis loads the slot map via
//     CLUSTER SLOTS), routed keyed ops (the client routes/follows MOVED), read-back, a cross-slot
//     multi-key op, and a hash-tag co-located multi-key op.
//
// Each checked behavior prints exactly one line:
//     RESULT ioredis <single|cluster> <op-group> <PASS|FAIL> [detail]
// The process exits non-zero if any group FAILed. Assertions check VALUES, not just "no throw".

"use strict";

const Redis = require("ioredis");

let failed = 0;

function result(mode, group, ok, detail = "") {
  const status = ok ? "PASS" : "FAIL";
  if (!ok) failed += 1;
  console.log(`RESULT ioredis ${mode} ${group} ${status} ${detail}`.trimEnd());
}

async function check(mode, group, fn) {
  try {
    const [ok, detail] = await fn();
    result(mode, group, ok, detail);
  } catch (err) {
    result(mode, group, false, `exception: ${err && err.message ? err.message : err}`);
  }
}

function parseArgs() {
  const args = { singlePort: 0, cluster: "", aclUser: "", aclPass: "", shardOwners: "" };
  const a = process.argv.slice(2);
  for (let i = 0; i < a.length; i++) {
    if (a[i] === "--single-port") args.singlePort = parseInt(a[++i], 10);
    else if (a[i] === "--cluster") args.cluster = a[++i];
    else if (a[i] === "--acl-user") args.aclUser = a[++i];
    else if (a[i] === "--acl-pass") args.aclPass = a[++i];
    else if (a[i] === "--shard-owners") args.shardOwners = a[++i];
  }
  return args;
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function runSingle(port) {
  const mode = "single";
  const pfx = `js:${Date.now() % 1000000}:`;
  const r = new Redis({ port, host: "127.0.0.1", lazyConnect: false });

  await check(mode, "connect", async () => {
    const pong = await r.ping();
    return [pong === "PONG", `PING=${pong}`];
  });

  await check(mode, "strings", async () => {
    const k = pfx + "str";
    await r.set(k, "hello");
    const v = await r.get(k);
    await r.append(k, " world");
    const full = await r.get(k);
    const rng = await r.getrange(k, 0, 4);
    await r.set(pfx + "n", "10");
    const n = await r.incrby(pfx + "n", 5);
    const ok = v === "hello" && full === "hello world" && rng === "hello" && n === 15;
    return [ok, `get=${v} append=${full} getrange=${rng} incr=${n}`];
  });

  await check(mode, "lists", async () => {
    const k = pfx + "list";
    await r.rpush(k, "a", "b", "c");
    await r.lpush(k, "z");
    const vals = await r.lrange(k, 0, -1);
    const ok = JSON.stringify(vals) === JSON.stringify(["z", "a", "b", "c"]);
    return [ok, `lrange=${JSON.stringify(vals)}`];
  });

  await check(mode, "hashes", async () => {
    const k = pfx + "hash";
    await r.hset(k, "f1", "v1", "f2", "v2");
    const m = await r.hgetall(k);
    const ok = m.f1 === "v1" && m.f2 === "v2" && Object.keys(m).length === 2;
    return [ok, `hgetall=${JSON.stringify(m)}`];
  });

  await check(mode, "sets", async () => {
    const k = pfx + "set";
    await r.sadd(k, "x", "y", "z");
    const members = (await r.smembers(k)).sort();
    const ok = JSON.stringify(members) === JSON.stringify(["x", "y", "z"]);
    return [ok, `smembers=${JSON.stringify(members)}`];
  });

  await check(mode, "zsets", async () => {
    const k = pfx + "zset";
    await r.zadd(k, 1, "one", 2, "two", 3, "three");
    const flat = await r.zrange(k, 0, -1, "WITHSCORES");
    // flat = [member, score, member, score, ...]
    const ok =
      flat[0] === "one" && flat[1] === "1" &&
      flat[2] === "two" && flat[3] === "2" &&
      flat[4] === "three" && flat[5] === "3";
    return [ok, `zrange_withscores=${JSON.stringify(flat)}`];
  });

  await check(mode, "expire-ttl", async () => {
    const k = pfx + "ttl";
    await r.set(k, "v");
    await r.expire(k, 100);
    const t = await r.ttl(k);
    const ok = t > 90 && t <= 100;
    return [ok, `ttl=${t}`];
  });

  await check(mode, "mget-mset", async () => {
    await r.mset(pfx + "m1", "A", pfx + "m2", "B", pfx + "m3", "C");
    const vals = await r.mget(pfx + "m1", pfx + "m2", pfx + "m3");
    const ok = JSON.stringify(vals) === JSON.stringify(["A", "B", "C"]);
    return [ok, `mget=${JSON.stringify(vals)}`];
  });

  await check(mode, "pipeline", async () => {
    const res = await r
      .pipeline()
      .set(pfx + "p1", "1")
      .incr(pfx + "p1")
      .get(pfx + "p1")
      .rpush(pfx + "pl", "a", "b")
      .lrange(pfx + "pl", 0, -1)
      .exec();
    // res = [[err, val], ...] in order
    const ok = res[1][1] === 2 && res[2][1] === "2" && JSON.stringify(res[4][1]) === JSON.stringify(["a", "b"]);
    return [ok, `pipeline=${JSON.stringify(res.map((x) => x[1]))}`];
  });

  await check(mode, "multi-exec", async () => {
    const res = await r.multi().set(pfx + "t", "0").incr(pfx + "t").incr(pfx + "t").exec();
    const final = await r.get(pfx + "t");
    const ok = res[res.length - 1][1] === 2 && final === "2";
    return [ok, `exec=${JSON.stringify(res.map((x) => x[1]))} final=${final}`];
  });

  await check(mode, "pubsub", async () => {
    const chan = pfx + "chan";
    const sub = new Redis({ port, host: "127.0.0.1" });
    const got = new Promise((resolve) => {
      sub.on("message", (c, m) => {
        if (c === chan) resolve(m);
      });
    });
    await sub.subscribe(chan);
    await sleep(100);
    await r.publish(chan, "ping-payload");
    const payload = await Promise.race([got, sleep(3000).then(() => null)]);
    await sub.quit();
    return [payload === "ping-payload", `payload=${payload}`];
  });

  await check(mode, "resp3", async () => {
    // CLIENT LIMITATION (documented finding): ioredis (v5) is RESP2-only. Its bundled `redis-parser`
    // has no case for the RESP3 map type byte `%` (nor `~`/`>`), so `HELLO 3` -- which IronCache
    // correctly answers with a RESP3 map -- makes the parser throw
    // `Protocol error, got "%" as reply type byte`. IronCache is NOT at fault here (redis-py and
    // go-redis both negotiate RESP3 against the SAME server cleanly); the client cannot consume
    // RESP3. We assert HELLO 2 works (the protocol ioredis supports) and record the RESP3 gap as a
    // KNOWN client limitation rather than an IronCache defect.
    const hello2 = await r.call("HELLO", "2");
    let proto = null;
    for (let i = 0; i + 1 < hello2.length; i += 2) {
      if (String(hello2[i]) === "proto") proto = Number(hello2[i + 1]);
    }
    const ok = proto === 2;
    return [ok, `hello2_proto=${proto} (RESP3 N/A: ioredis is RESP2-only -- client limitation, not IronCache)`];
  });

  r.disconnect();
}

// runCluster drives the ioredis cluster client against a set of slot-owner endpoints. The SAME body
// serves TWO legs, picked by `mode`: the turnkey 3-node Raft cluster (mode="cluster") and the #517
// single-node SHARD-OWNERS projection (mode="shard-owners"), which exposes the node's N internal
// shards as N slot owners on distinct ports. In both, a cluster-aware client discovers the topology
// via CLUSTER SLOTS and routes each key to its owner by following MOVED; for shard-owners that owner
// is the key's home SHARD's port, so the internal cross-shard hop is eliminated (the zero-hop metric
// is asserted by run.sh + the Rust metrics_endpoint.rs test).
async function runCluster(seeds, mode = "cluster") {
  const pfx = `js-${mode}:${Date.now() % 1000000}:`;
  const nodes = seeds.map((s) => {
    const [host, port] = s.split(":");
    return { host, port: parseInt(port, 10) };
  });

  const cluster = new Redis.Cluster(nodes, {
    // IronCache advertises 127.0.0.1; no NAT mapping needed.
    redisOptions: {},
    clusterRetryStrategy: (times) => (times > 5 ? null : 200),
  });

  // DISCOVERY: ioredis loads the slot map via CLUSTER SLOTS once connected.
  let discovered = false;
  await check(mode, "discovery", async () => {
    await new Promise((resolve, reject) => {
      const t = setTimeout(() => reject(new Error("cluster ready timeout")), 15000);
      cluster.on("ready", () => {
        clearTimeout(t);
        resolve();
      });
      cluster.on("error", () => {}); // transient connect errors during discovery are expected
    });
    const pong = await cluster.ping();
    const nodeCount = cluster.nodes("master").length;
    discovered = pong === "PONG" && nodeCount >= 1;
    return [discovered, `masters_discovered=${nodeCount}`];
  });

  if (!discovered) {
    for (const g of ["routed-ops", "routed-readback", "crossslot", "hashtag-coloc", "pipeline"]) {
      result(mode, g, false, "discovery failed");
    }
    cluster.disconnect();
    return;
  }

  await check(mode, "routed-ops", async () => {
    for (let i = 0; i < 60; i++) {
      await cluster.set(`${pfx}k${i}`, String(i));
    }
    return [true, "set 60 keys across slots"];
  });

  await check(mode, "routed-readback", async () => {
    for (let i = 0; i < 60; i++) {
      const v = await cluster.get(`${pfx}k${i}`);
      if (v !== String(i)) return [false, `k${i}=${v}`];
    }
    return [true, "all 60 read back"];
  });

  await check(mode, "crossslot", async () => {
    try {
      await cluster.mget(`${pfx}k0`, `${pfx}k1`, `${pfx}k2`);
      return [true, "mget across slots accepted (client may split)"];
    } catch (err) {
      const msg = (err.message || String(err)).toUpperCase();
      const ok = msg.includes("CROSSSLOT") || msg.includes("SLOT");
      return [ok, `rejected: ${err.message}`];
    }
  });

  await check(mode, "hashtag-coloc", async () => {
    await cluster.mset(`${pfx}{t}:a`, "1", `${pfx}{t}:b`, "2", `${pfx}{t}:c`, "3");
    const vals = await cluster.mget(`${pfx}{t}:a`, `${pfx}{t}:b`, `${pfx}{t}:c`);
    const ok = JSON.stringify(vals) === JSON.stringify(["1", "2", "3"]);
    return [ok, `hashtag_mget=${JSON.stringify(vals)}`];
  });

  await check(mode, "pipeline", async () => {
    // A cluster pipeline must be slot-coherent in ioredis; use a hash tag so all keys co-locate.
    const p = cluster.pipeline();
    for (let i = 0; i < 10; i++) p.set(`${pfx}{pp}:${i}`, `val${i}`);
    await p.exec();
    for (let i = 0; i < 10; i++) {
      const v = await cluster.get(`${pfx}{pp}:${i}`);
      if (v !== `val${i}`) return [false, `pp${i}=${v}`];
    }
    return [true, "cluster pipeline 10 co-located keys ok"];
  });

  cluster.disconnect();
}

async function runRestricted(seeds, user, pass) {
  // #405 per-subcommand-ACL leg: drive the cluster as the LOCKED-DOWN `svc` user
  // (`+@read +@write +@connection +@transaction -@dangerous +cluster|slots|shards|nodes|info`).
  // Proves (a) a scoped user can DISCOVER + do a routed SET/GET round-trip and (b) every CLUSTER
  // MUTATOR is NOPERM (CLUSTER ADDSLOTS denied -> the group PASSES). NOTE: ioredis's default
  // enableReadyCheck issues `CLUSTER INFO`, which is WHY the svc grant includes `+cluster|info`
  // (go-redis / redis-py discover via CLUSTER SLOTS alone) -- the documented #405 finding.
  const mode = "restricted";
  const pfx = `jsr:${Date.now() % 1000000}:`;
  const nodes = seeds.map((s) => {
    const [host, port] = s.split(":");
    return { host, port: parseInt(port, 10) };
  });

  const cluster = new Redis.Cluster(nodes, {
    redisOptions: { username: user, password: pass },
    clusterRetryStrategy: (times) => (times > 5 ? null : 200),
  });

  let discovered = false;
  await check(mode, "discovery", async () => {
    await new Promise((resolve, reject) => {
      const t = setTimeout(() => reject(new Error("cluster ready timeout")), 15000);
      cluster.on("ready", () => {
        clearTimeout(t);
        resolve();
      });
      cluster.on("error", () => {}); // transient connect errors during discovery are expected
    });
    const pong = await cluster.ping();
    const nodeCount = cluster.nodes("master").length;
    discovered = pong === "PONG" && nodeCount >= 1;
    return [discovered, `masters_discovered=${nodeCount} as ${user}`];
  });

  if (!discovered) {
    for (const g of ["rw-roundtrip", "addslots-denied"]) {
      result(mode, g, false, "discovery failed");
    }
    cluster.disconnect();
    return;
  }

  await check(mode, "rw-roundtrip", async () => {
    for (let i = 0; i < 30; i++) await cluster.set(`${pfx}k${i}`, String(i));
    for (let i = 0; i < 30; i++) {
      const v = await cluster.get(`${pfx}k${i}`);
      if (v !== String(i)) return [false, `mismatch k${i}=${v}`];
    }
    return [true, "set+get 30 keys across slots as svc"];
  });

  // SECURITY BOUNDARY: a CLUSTER MUTATOR must be NOPERM for svc. Use a DIRECT (non-cluster)
  // connection to node[0] as svc for a deterministic target; PASS iff ADDSLOTS is denied NOPERM.
  await check(mode, "addslots-denied", async () => {
    const one = new Redis({ host: nodes[0].host, port: nodes[0].port, username: user, password: pass });
    try {
      await one.call("CLUSTER", "ADDSLOTS", "0");
      return [false, "CLUSTER ADDSLOTS was ACCEPTED for a -@dangerous user (escalation!)"];
    } catch (err) {
      const msg = (err && err.message ? err.message : String(err)).toUpperCase();
      return [msg.includes("NOPERM"), `denied: ${err && err.message ? err.message : err}`];
    } finally {
      one.disconnect();
    }
  });

  cluster.disconnect();
}

async function main() {
  const args = parseArgs();
  console.log(`# ioredis ${require("ioredis/package.json").version}`);
  if (args.singlePort > 0) await runSingle(args.singlePort);
  if (args.cluster) await runCluster(args.cluster.split(","));
  if (args.cluster && args.aclUser) await runRestricted(args.cluster.split(","), args.aclUser, args.aclPass);
  if (args.shardOwners) await runCluster(args.shardOwners.split(","), "shard-owners");
  process.exit(failed > 0 ? 1 : 0);
}

main().catch((err) => {
  console.error(err);
  process.exit(2);
});

File: tests/drivers/python/driver_compat.py
# SPDX-License-Identifier: MIT OR Apache-2.0
"""PROD-8 redis-py compatibility driver (issue #158).

Runs the real `redis` Python client against a live IronCache, in TWO modes:

  * single-node (redis.Redis): core data-type ops, pipelining, MULTI/EXEC, pub/sub, RESP3 (HELLO 3 +
    a push round-trip), and -- via a second connection that sets requirepass through CONFIG-less env
    is out of scope here, so AUTH is exercised only if IRONCACHE_REQUIREPASS was set for the server.
  * cluster (redis.cluster.RedisCluster): topology DISCOVERY via CLUSTER SLOTS, routed keyed ops
    (the client must FOLLOW MOVED to reach the owning node), read-back correctness, and a CROSSSLOT
    multi-key op (the client/server must reject a multi-key op that spans slots).

Each checked behavior prints exactly one line:
    RESULT redis-py <single|cluster> <op-group> <PASS|FAIL> [detail]
The process exits non-zero if any group FAILed (a harness signal; the orchestrator still collects
every RESULT line). This asserts VALUES, not just "no exception": e.g. ZRANGE withscores returns the
right (member, score) pairs, HGETALL the right dict, a pipeline returns replies in order, MULTI/EXEC
is atomic, a SUBSCRIBE receives a PUBLISH, and HELLO 3 returns a server map.
"""

from __future__ import annotations

import argparse
import sys
import threading
import time
import uuid

import redis


FAILED = 0


def result(mode: str, group: str, ok: bool, detail: str = "") -> None:
    global FAILED
    status = "PASS" if ok else "FAIL"
    if not ok:
        FAILED += 1
    print(f"RESULT redis-py {mode} {group} {status} {detail}".rstrip(), flush=True)


def check(mode: str, group: str, fn) -> None:
    """Run one op-group closure; PASS iff it returns truthy and raises nothing."""
    try:
        ok, detail = fn()
    except Exception as exc:  # a client-side exception IS the finding
        result(mode, group, False, f"exception: {type(exc).__name__}: {exc}")
        return
    result(mode, group, bool(ok), detail)


# --------------------------------------------------------------------------- single-node groups
def run_single(port: int) -> None:
    mode = "single"
    pfx = f"py:{uuid.uuid4().hex[:8]}:"
    r = redis.Redis(host="127.0.0.1", port=port, decode_responses=True)

    def g_connect():
        return r.ping() is True, "PING"

    def g_strings():
        k = pfx + "str"
        r.set(k, "hello")
        v = r.get(k)
        r.append(k, " world")
        full = r.get(k)
        rng = r.getrange(k, 0, 4)
        r.set(pfx + "n", "10")
        n = r.incr(pfx + "n", 5)
        ok = v == "hello" and full == "hello world" and rng == "hello" and n == 15
        return ok, f"get={v!r} append={full!r} getrange={rng!r} incr={n}"

    def g_lists():
        k = pfx + "list"
        r.rpush(k, "a", "b", "c")
        r.lpush(k, "z")
        vals = r.lrange(k, 0, -1)
        ok = vals == ["z", "a", "b", "c"]
        return ok, f"lrange={vals}"

    def g_hashes():
        k = pfx + "hash"
        r.hset(k, mapping={"f1": "v1", "f2": "v2"})
        m = r.hgetall(k)
        ok = m == {"f1": "v1", "f2": "v2"}
        return ok, f"hgetall={m}"

    def g_sets():
        k = pfx + "set"
        r.sadd(k, "x", "y", "z")
        members = set(r.smembers(k))
        ok = members == {"x", "y", "z"}
        return ok, f"smembers={sorted(members)}"

    def g_zsets():
        k = pfx + "zset"
        r.zadd(k, {"one": 1.0, "two": 2.0, "three": 3.0})
        pairs = r.zrange(k, 0, -1, withscores=True)
        ok = pairs == [("one", 1.0), ("two", 2.0), ("three", 3.0)]
        return ok, f"zrange_withscores={pairs}"

    def g_ttl():
        k = pfx + "ttl"
        r.set(k, "v")
        r.expire(k, 100)
        t = r.ttl(k)
        ok = 90 <= t <= 100
        return ok, f"ttl={t}"

    def g_mget_mset():
        r.mset({pfx + "m1": "A", pfx + "m2": "B", pfx + "m3": "C"})
        vals = r.mget(pfx + "m1", pfx + "m2", pfx + "m3")
        ok = vals == ["A", "B", "C"]
        return ok, f"mget={vals}"

    def g_pipeline():
        p = r.pipeline(transaction=False)
        p.set(pfx + "p1", "1")
        p.incr(pfx + "p1")
        p.get(pfx + "p1")
        p.lpush(pfx + "pl", "a", "b")
        p.lrange(pfx + "pl", 0, -1)
        res = p.execute()
        ok = res[1] == 2 and res[2] == "2" and res[4] == ["b", "a"]
        return ok, f"pipeline_replies={res}"

    def g_multi_exec():
        # MULTI/EXEC atomic transaction (transaction=True is the default pipeline). The single node
        # is booted with one shard (see run.sh) so all keys share a home shard: IronCache does not
        # support cross-shard transactions, so a multi-shard node would abort a txn whose keys land
        # on a non-home shard (a documented model detail, see DRIVER_MATRIX.md).
        tk = pfx + "t"
        with r.pipeline(transaction=True) as p:
            p.set(tk, "0")
            p.incr(tk)
            p.incr(tk)
            res = p.execute()
        final = r.get(tk)
        ok = res[-1] == 2 and final == "2"
        return ok, f"exec={res} final={final!r}"

    def g_pubsub():
        chan = pfx + "chan"
        sub = redis.Redis(host="127.0.0.1", port=port, decode_responses=True)
        ps = sub.pubsub()
        ps.subscribe(chan)
        # Drain the subscribe-confirmation message.
        deadline = time.time() + 3
        got_sub = False
        while time.time() < deadline:
            m = ps.get_message(timeout=0.5)
            if m and m.get("type") == "subscribe":
                got_sub = True
                break
        # Publish from the main connection; the subscriber must receive it.
        time.sleep(0.1)
        r.publish(chan, "ping-payload")
        payload = None
        deadline = time.time() + 3
        while time.time() < deadline:
            m = ps.get_message(timeout=0.5)
            if m and m.get("type") == "message":
                payload = m.get("data")
                break
        ps.close()
        sub.close()
        ok = got_sub and payload == "ping-payload"
        return ok, f"subscribed={got_sub} payload={payload!r}"

    def g_resp3():
        # RESP3: HELLO 3 negotiates protocol 3; server returns a map. Then a normal op still works,
        # and a push-delivering pub/sub round-trip works under RESP3.
        r3 = redis.Redis(host="127.0.0.1", port=port, decode_responses=True, protocol=3)
        hello = r3.execute_command("HELLO", "3")
        proto = hello.get("proto") if isinstance(hello, dict) else None
        r3.set(pfx + "r3", "ok")
        v = r3.get(pfx + "r3")
        # RESP3 push: subscribe + publish on RESP3 connections.
        chan = pfx + "r3chan"
        sub = redis.Redis(host="127.0.0.1", port=port, decode_responses=True, protocol=3)
        ps = sub.pubsub()
        ps.subscribe(chan)
        deadline = time.time() + 3
        while time.time() < deadline:
            m = ps.get_message(timeout=0.5)
            if m and m.get("type") == "subscribe":
                break
        time.sleep(0.1)
        r3.publish(chan, "r3-push")
        payload = None
        deadline = time.time() + 3
        while time.time() < deadline:
            m = ps.get_message(timeout=0.5)
            if m and m.get("type") == "message":
                payload = m.get("data")
                break
        ps.close()
        sub.close()
        r3.close()
        ok = proto == 3 and v == "ok" and payload == "r3-push"
        return ok, f"hello_proto={proto} get={v!r} push={payload!r}"

    check(mode, "connect", g_connect)
    check(mode, "strings", g_strings)
    check(mode, "lists", g_lists)
    check(mode, "hashes", g_hashes)
    check(mode, "sets", g_sets)
    check(mode, "zsets", g_zsets)
    check(mode, "expire-ttl", g_ttl)
    check(mode, "mget-mset", g_mget_mset)
    check(mode, "pipeline", g_pipeline)
    check(mode, "multi-exec", g_multi_exec)
    check(mode, "pubsub", g_pubsub)
    check(mode, "resp3", g_resp3)
    r.close()


# --------------------------------------------------------------------------- cluster groups
def run_cluster(nodes: list[tuple[str, int]], mode: str = "cluster") -> None:
    # `mode` labels the RESULT lines and picks the key prefix. The SAME cluster-client body drives
    # the turnkey 3-node Raft cluster (`mode="cluster"`) AND the #517 single-node SHARD-OWNERS
    # projection (`mode="shard-owners"`): both expose N slot owners on distinct endpoints that a
    # cluster-aware client discovers via CLUSTER SLOTS and routes to by following MOVED. The
    # shard-owners leg proves a real client routes each key to its owning SHARD's port, so the
    # internal cross-shard hop is eliminated (the zero-hop metric is asserted by run.sh + the Rust
    # metrics_endpoint.rs test).
    from redis.cluster import ClusterNode, RedisCluster

    pfx = f"py:{mode}:{uuid.uuid4().hex[:8]}:"
    startup = [ClusterNode(h, p) for (h, p) in nodes]

    # Construction itself exercises DISCOVERY: redis-py issues CLUSTER SLOTS against a seed and builds
    # the slot->node map. A malformed CLUSTER SLOTS reply would raise here.
    try:
        rc = RedisCluster(startup_nodes=startup, decode_responses=True, require_full_coverage=True)
        rc.ping()
        discovered = True
        ndetail = f"nodes_discovered={len(rc.get_nodes())}"
    except Exception as exc:
        result(mode, "discovery", False, f"exception: {type(exc).__name__}: {exc}")
        # Without discovery the rest is moot; record them as FAIL with the cause.
        for g in ("routed-ops", "routed-readback", "crossslot", "pipeline"):
            result(mode, g, False, "discovery failed")
        return
    result(mode, "discovery", discovered, ndetail)

    def g_routed_ops():
        # Write keys whose slots land in each of the three blocks; the client must route each to its
        # owner by FOLLOWING MOVED (the seed node only owns one third). Use distinct keys.
        wrote = 0
        for i in range(60):
            rc.set(f"{pfx}k{i}", str(i))
            wrote += 1
        return wrote == 60, f"set_count={wrote}"

    def g_routed_readback():
        ok = True
        bad = ""
        for i in range(60):
            v = rc.get(f"{pfx}k{i}")
            if v != str(i):
                ok = False
                bad = f"k{i}={v!r}"
                break
        return ok, ("all 60 read back" if ok else f"mismatch {bad}")

    def g_crossslot():
        # A multi-key op whose keys hash to DIFFERENT slots must be rejected. redis-py's cluster
        # client rejects this CLIENT-SIDE (it computes both slots from the COMMAND key-spec the
        # server reported, sees they differ, and raises RedisClusterException) -- which proves the
        # client successfully learned IronCache's command table + slot function. Some redis-py
        # versions instead surface the server's -CROSSSLOT as a ResponseError. Both are correct.
        from redis.exceptions import RedisError, RedisClusterException
        try:
            rc.mget(f"{pfx}k0", f"{pfx}k1", f"{pfx}k2")
            # No raise means the client split the op per-slot client-side; also acceptable.
            return True, "mget across slots handled (client-side split or accepted)"
        except (RedisError, RedisClusterException) as exc:
            msg = str(exc)
            ok = "CROSSSLOT" in msg.upper() or "slot" in msg.lower()
            return ok, f"rejected: {msg[:90]}"

    def g_pipeline():
        # Cluster pipeline: redis-py routes each command to its owner node and gathers replies.
        with rc.pipeline() as p:
            for i in range(10):
                p.set(f"{pfx}pp{i}", f"val{i}")
            p.execute()
        vals = [rc.get(f"{pfx}pp{i}") for i in range(10)]
        ok = vals == [f"val{i}" for i in range(10)]
        return ok, f"pipeline_readback_ok={ok}"

    def g_hashtag():
        # Hash-tagged keys force co-location in ONE slot, so a multi-key op succeeds. Proves the
        # client respects {tag} slot computation against IronCache's CRC16.
        rc.mset({f"{pfx}{{t}}:a": "1", f"{pfx}{{t}}:b": "2", f"{pfx}{{t}}:c": "3"})
        vals = rc.mget(f"{pfx}{{t}}:a", f"{pfx}{{t}}:b", f"{pfx}{{t}}:c")
        ok = vals == ["1", "2", "3"]
        return ok, f"hashtag_mget={vals}"

    check(mode, "routed-ops", g_routed_ops)
    check(mode, "routed-readback", g_routed_readback)
    check(mode, "crossslot", g_crossslot)
    check(mode, "pipeline", g_pipeline)
    check(mode, "hashtag-coloc", g_hashtag)
    try:
        rc.close()
    except Exception:
        pass


# --------------------------------------------------------------------------- restricted-user groups
def run_restricted(nodes: list[tuple[str, int]], user: str, password: str) -> None:
    """#405 per-subcommand-ACL leg: drive the cluster as the LOCKED-DOWN `svc` user
    (`+@read +@write +@connection +@transaction -@dangerous +cluster|slots|shards|nodes|info`).

    Proves (a) a scoped user can still DISCOVER topology + do a routed SET/GET round-trip, and
    (b) every CLUSTER MUTATOR is NOPERM -- CLUSTER ADDSLOTS must be denied (the group PASSES when it
    is). redis-py applies username/password to every node connection, so MOVED-routing re-AUTHs as
    `svc` transparently.
    """
    mode = "restricted"
    from redis.cluster import ClusterNode, RedisCluster
    from redis.exceptions import ResponseError

    pfx = f"pyr:{uuid.uuid4().hex[:8]}:"
    startup = [ClusterNode(h, p) for (h, p) in nodes]

    # DISCOVERY as svc: construction issues CLUSTER SLOTS (granted). A NOPERM would raise here.
    try:
        rc = RedisCluster(
            startup_nodes=startup,
            decode_responses=True,
            require_full_coverage=True,
            username=user,
            password=password,
        )
        rc.ping()
        result(mode, "discovery", True, f"nodes_discovered={len(rc.get_nodes())} as {user}")
    except Exception as exc:
        result(mode, "discovery", False, f"exception: {type(exc).__name__}: {exc}")
        for g in ("rw-roundtrip", "addslots-denied"):
            result(mode, g, False, "discovery failed")
        return

    def g_rw_roundtrip():
        for i in range(30):
            rc.set(f"{pfx}k{i}", str(i))
        for i in range(30):
            v = rc.get(f"{pfx}k{i}")
            if v != str(i):
                return False, f"mismatch k{i}={v!r}"
        return True, "set+get 30 keys across slots as svc"

    def g_addslots_denied():
        # A CLUSTER MUTATOR must be NOPERM for svc. Use a DIRECT (non-cluster) connection to node[0]
        # authenticated as svc for a deterministic target; PASS iff ADDSLOTS is denied NOPERM.
        host, port = nodes[0]
        one = redis.Redis(host=host, port=port, decode_responses=True, username=user, password=password)
        try:
            one.execute_command("CLUSTER", "ADDSLOTS", 0)
            return False, "CLUSTER ADDSLOTS was ACCEPTED for a -@dangerous user (escalation!)"
        except ResponseError as exc:
            # redis-py raises NoPermissionError and STRIPS the NOPERM error-code
            # prefix, so its message is "... has no permissions ..." without the
            # literal token (go-redis / ioredis surface the raw NOPERM). Accept
            # either form: the command was denied as long as it is a permissions
            # error and not some unrelated ResponseError.
            msg = str(exc)
            denied = "NOPERM" in msg.upper() or "NO PERMISSIONS" in msg.upper()
            return denied, f"denied: {msg[:90]}"
        finally:
            one.close()

    check(mode, "rw-roundtrip", g_rw_roundtrip)
    check(mode, "addslots-denied", g_addslots_denied)
    try:
        rc.close()
    except Exception:
        pass


def parse_nodes(csv: str) -> list[tuple[str, int]]:
    out = []
    for part in csv.split(","):
        host, _, port = part.strip().rpartition(":")
        out.append((host, int(port)))
    return out


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--single-port", type=int, required=True)
    ap.add_argument("--cluster", type=str, required=True, help="comma list of host:port seeds")
    ap.add_argument("--acl-user", type=str, default="", help="restricted (scoped-ACL) username, #405 leg")
    ap.add_argument("--acl-pass", type=str, default="", help="restricted (scoped-ACL) password, #405 leg")
    ap.add_argument("--shard-owners", type=str, default="",
                    help="comma list of the N shard-owner host:port endpoints (#517 leg)")
    args = ap.parse_args()

    print(f"# redis-py {redis.__version__}", flush=True)
    run_single(args.single_port)
    run_cluster(parse_nodes(args.cluster))
    if args.acl_user:
        run_restricted(parse_nodes(args.cluster), args.acl_user, args.acl_pass)
    if args.shard_owners:
        run_cluster(parse_nodes(args.shard_owners), mode="shard-owners")
    return 1 if FAILED else 0


if __name__ == "__main__":
    sys.exit(main())
